File: [iOS] HEALTH 1ST/maps.ios-master/NMapSampleObjc/NMapSampleObjc/RotationViewController.m
//
//  RotationViewController.m
//  NMapSampleObjc
//
//  Created by Naver on 2016. 11. 21..
//  Copyright © 2016년 Naver. All rights reserved.
//

#import "RotationViewController.h"
#import "NMapViewResources.h"

@interface RotationViewController ()

@property (nonatomic, strong) NMapView * mapView;
@property (nonatomic) CGFloat viewAngle;

- (IBAction)leftButtonClicked:(id)sender;
- (IBAction)rightButtonClicked:(id)sender;

@end

@implementation RotationViewController

BOOL hasInit = NO;

- (void)viewDidLoad {
    [super viewDidLoad];
    self.mapView = [[NMapView alloc] initWithFrame:self.view.frame];
    self.navigationController.navigationBar.translucent = NO;

    // set the delegate for map view
    [self.mapView setDelegate:self];

    // set the application api key for Open MapViewer Library
    [self.mapView setClientId:@"YOUR CLIENT ID"];
 
    self.mapView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
   
    [self.view insertSubview:self.mapView atIndex:0];
}

- (void) viewDidLayoutSubviews {
    [super viewDidLayoutSubviews];
    [self.mapView setFrame:self.view.bounds];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [self.mapView viewWillAppear];
}

- (void)viewDidAppear:(BOOL)animated {
    [self.mapView viewDidAppear];
    [super viewDidAppear:animated];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    [self.mapView viewWillDisappear];
}

- (void)viewDidDisappear:(BOOL)animated {
    [self.mapView viewDidDisappear];
    [super viewDidDisappear:animated];
}

- (void) viewWillTransitionToSize:(CGSize)size withTransitionCoordinator:(id<UIViewControllerTransitionCoordinator>)coordinator {
    
    if (size.width != self.view.frame.size.width) {
        if ([self.mapView isAutoRotateEnabled] == YES && hasInit == YES) {
            self.viewAngle = self.mapView.rotateAngle;
            [self.mapView setAutoRotateEnabled:NO animate:NO];
            [coordinator animateAlongsideTransition:^ (id <UIViewControllerTransitionCoordinatorContext> context) {
                if (hasInit == YES) {
                    [self.mapView setAutoRotateEnabled:YES animate:YES];
                    [self.mapView setRotateAngle:self.viewAngle];
                }
        
            } completion:nil];
        }
    }
    
}

#pragma mark NMapViewDelegate Method

- (void) onMapView:(NMapView *)mapView initHandler:(NMapError *)error {
    if (error == nil) { // success
        // set map center and level
        [self.mapView setMapCenter:NGeoPointMake(126.978371, 37.5666091) atLevel:11];
        // set for retina display
        [self.mapView setMapEnlarged:YES mapHD:YES];
        // set default map mode
        [self.mapView setMapViewMode:NMapViewModeVector];
        hasInit = YES;
    } else { // fail
        NSLog(@"onMapView:initHandler: %@", [error description]);
    }
}

#pragma mark NMapPOIdataOverlayDelegate


- (UIImage *) onMapOverlay:(NMapPOIdataOverlay *)poiDataOverlay imageForOverlayItem:(NMapPOIitem *)poiItem selected:(BOOL)selected {
    return [NMapViewResources imageWithType:poiItem.poiFlagType selected:selected];
}

- (CGPoint) onMapOverlay:(NMapPOIdataOverlay *)poiDataOverlay anchorPointWithType:(NMapPOIflagType)poiFlagTyp{
    return [NMapViewResources anchorPoint:poiFlagTyp];
}

- (UIImage*) onMapOverlay:(NMapPOIdataOverlay *)poiDataOverlay imageForCalloutOverlayItem:(NMapPOIitem *)poiItem
           constraintSize:(CGSize)constraintSize selected:(BOOL)selected
imageForCalloutRightAccessory:(UIImage *)imageForCalloutRightAccessory
          calloutPosition:(CGPoint *)calloutPosition calloutHitRect:(CGRect *)calloutHitRect {
    return nil;
}

- (CGPoint) onMapOverlay:(NMapPOIdataOverlay *)poiDataOverlay calloutOffsetWithType:(NMapPOIflagType)poiFlagType {
    return CGPointZero;
}

#pragma mark RotationViewController

- (IBAction)leftButtonClicked:(id)sender {
    if (self.mapView.isAutoRotateEnabled == NO) {
        [self.mapView setAutoRotateEnabled:YES animate:YES];
    }
    self.viewAngle += 45.f;
    [self.mapView setRotateAngle:self.viewAngle];
}

- (IBAction)rightButtonClicked:(id)sender {
    if (self.mapView.isAutoRotateEnabled == NO) {
        [self.mapView setAutoRotateEnabled:YES animate:YES];
    }
    self.viewAngle -= 45.f;
    [self.mapView setRotateAngle:self.viewAngle];
}
@end

File: [iOS] HEALTH 1ST/maps.ios-master/NMapSampleSwift/NMapSampleSwift/RotationViewController.swift
//
//  RotationViewController.swift
//  NMapSampleSwift
//
//  Created by Junggyun Ahn on 2016. 11. 22..
//  Copyright © 2016년 Naver. All rights reserved.
//

class RotationViewController: UIViewController, NMapViewDelegate, NMapPOIdataOverlayDelegate {
    
    let rotateAnglePerClick:CGFloat = 45
    
    var mapView: NMapView?
    
    var previousAngle:CGFloat = 0
    
    var hasInit = false
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.navigationController?.navigationBar.isTranslucent = false
        
        mapView = NMapView(frame: self.view.frame)
        
        if let mapView = mapView {
            
            // set the delegate for map view
            mapView.delegate = self
            
            // set the application api key for Open MapViewer Library
            mapView.setClientId("YOUR CLIENT ID")
            
            mapView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
            
            view.insertSubview(mapView, at: 0)
        }        
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        mapView?.didReceiveMemoryWarning()
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        mapView?.viewWillAppear()
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        mapView?.viewDidAppear()
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        mapView?.viewWillDisappear()
    }
    
    override func viewDidDisappear(_ animated: Bool) {
        super.viewDidDisappear(animated)
        mapView?.viewDidDisappear()
    }
    
    override func viewWillTransition(to size: CGSize, with coordinator: UIViewControllerTransitionCoordinator) {
        super.viewWillTransition(to: size, with: coordinator)
        
        if view.frame.size.width != size.width {
            if let mapView = mapView, hasInit, mapView.isAutoRotateEnabled {
                previousAngle = mapView.rotateAngle
                mapView.setAutoRotateEnabled(false, animate: false)
                
                coordinator.animate(alongsideTransition: {(context: UIViewControllerTransitionCoordinatorContext) -> Void in
                    if let mapView = self.mapView, self.hasInit {
                        mapView.setAutoRotateEnabled(true, animate: false)
                        mapView.setRotateAngle(Float(self.previousAngle), animate: true)
                    }
                }, completion: nil)
            }
        }
    }
    
    // MARK: - NMapViewDelegate Methods
    
    open func onMapView(_ mapView: NMapView!, initHandler error: NMapError!) {
        if (error == nil) { // success
            // set map center and level
            mapView.setMapCenter(NGeoPoint(longitude:126.978371, latitude:37.5666091), atLevel:11)
            // set for retina display
            mapView.setMapEnlarged(true, mapHD: true)
            // set map mode : vector/satelite/hybrid
            mapView.mapViewMode = .vector
            
            hasInit = true
        } else { // fail
            print("onMapView:initHandler: \(error.description)")
        }
    }
    
    // MARK: - NMapPOIdataOverlayDelegate Methods
    
    open func onMapOverlay(_ poiDataOverlay: NMapPOIdataOverlay!, imageForOverlayItem poiItem: NMapPOIitem!, selected: Bool) -> UIImage! {
        return NMapViewResources.imageWithType(poiItem.poiFlagType, selected: selected);
    }
    
    open func onMapOverlay(_ poiDataOverlay: NMapPOIdataOverlay!, anchorPointWithType poiFlagType: NMapPOIflagType) -> CGPoint {
        return NMapViewResources.anchorPoint(withType: poiFlagType)
    }
    
    open func onMapOverlay(_ poiDataOverlay: NMapPOIdataOverlay!, calloutOffsetWithType poiFlagType: NMapPOIflagType) -> CGPoint {
        return CGPoint.zero
    }
    
    open func onMapOverlay(_ poiDataOverlay: NMapPOIdataOverlay!, imageForCalloutOverlayItem poiItem: NMapPOIitem!, constraintSize: CGSize, selected: Bool, imageForCalloutRightAccessory: UIImage!, calloutPosition: UnsafeMutablePointer<CGPoint>!, calloutHit calloutHitRect: UnsafeMutablePointer<CGRect>!) -> UIImage! {
        return nil
    }
    
    // MARK: - UI Actions
    
    @IBAction func rightButtonClicked(_ sender: UIButton) {
        if let mapView = mapView {
            
            if mapView.isAutoRotateEnabled == false {
                mapView.setAutoRotateEnabled(true, animate: true)
            }
            
            let newAngle = mapView.rotateAngle - rotateAnglePerClick// + (mapView.rotateAngle < rotateAnglePerClick ? 360 : 0)
            mapView.setRotateAngle(Float(newAngle), animate: true)
        }
    }
    
    @IBAction func leftButtonClicked(_ sender: UIButton) {
        if let mapView = mapView {
            
            if mapView.isAutoRotateEnabled == false {
                mapView.setAutoRotateEnabled(true, animate: true)
            }
            
            let newAngle = mapView.rotateAngle + rotateAnglePerClick// - (mapView.rotateAngle > 360 - rotateAnglePerClick ? 360 : 0)
            mapView.setRotateAngle(Float(newAngle), animate: true)
        }
    }
}
